lib/src/entrypoint.dart

```diff
@@ -500,50 +530,135 @@ class Entrypoint {
+            !fileExists(p.join(lockFilePackagePath, "pubspec.yaml"))) {
+          return false;
+        }
+
+        return true;
+      });
+
   /// Determines whether or not the `.packages` file is out of date with respect
-  /// to the lockfile.
+  /// to the [lockfile].
   ///
-  /// This will be `false` if the packages file contains dependencies that are
-  /// not in the lockfile or that don't match what's in there.
+  /// This will be `false` if [lockfile] contains dependencies that  are not in
+  /// the `.packages` or that don't match what's in there.
   bool _isPackagesFileUpToDate() {
     var packages = packages_file.parse(
         File(packagesFile).readAsBytesSync(), p.toUri(packagesFile));
 
-    return lockFile.packages.values.every((lockFileId) {
-      // It's very unlikely that the lockfile is invalid here, but it's not
-      // impossible—for example, the user may have a very old application
-      // package with a checked-in lockfile that's newer than the pubspec, but
-      // that contains SDK dependencies.
-      if (lockFileId.source is UnknownSource) return false;
-
-      var packagesFileUri = packages[lockFileId.name];
-      if (packagesFileUri == null) return false;
+    final packagePathsMapping = <String, String>{};
+    for (final package in packages.keys) {
+      final packageUri = packages[package];
 
       // Pub only generates "file:" and relative URIs.
-      if (packagesFileUri.scheme != 'file' &&
-          packagesFileUri.scheme.isNotEmpty) {
+      if (packageUri.scheme != 'file' && packageUri.scheme.isNotEmpty) {
         return false;
       }
 
-      var source = cache.source(lockFileId.source);
-
       // Get the dirname of the .packages path, since it's pointing to lib/.
-      var packagesFilePath =
-          p.dirname(p.join(root.dir, p.fromUri(packagesFileUri)));
-      var lockFilePath = p.join(root.dir, source.getDirectory(lockFileId));
-
-      // For cached sources, make sure the directory exists and looks like a
-      // package. This is also done by [_arePackagesAvailable] but that may not
-      // be run if the lockfile is newer than the pubspec.
-      if (source is CachedSource && !dirExists(packagesFilePath) ||
-          !fileExists(p.join(packagesFilePath, "pubspec.yaml"))) {
+      final packagePath = p.dirname(p.join(root.dir, p.fromUri(packageUri)));
+      packagePathsMapping[package] = packagePath;
+    }
+
+    return _isLockFileSatisfied(packagePathsMapping);
+  }
+
+  /// Determines whether or not the `.dart_tool/package_config.json` file is
+  /// out of date with respect to the lockfile.
+  ///
+  /// This will be `false` if the [lockfile] contains dependencies that are not
+  /// in the `.dart_tool/package_config.json` or that don't match what's in
+  /// there.
+  ///
+  /// Throws [DataException], if `.dart_tool/package_config.json` is not
+  /// up-to-date for some other reason.
+  bool _isPackageConfigUpToDate() {
```

Comment:
We should probably have return type `void` and just throw where we now return `false`